core/src/qos_service.rs

```diff
@@ -177,28 +177,58 @@ impl QosService {
     pub fn update_or_remove_transaction_costs<'a>(
         transaction_costs: impl Iterator<Item = &'a TransactionCost>,
         transaction_qos_results: impl Iterator<Item = &'a transaction::Result<()>>,
-        retryable_transaction_indexes: &[usize],
+        transaction_committed_status: Option<&Vec<bool>>,
+        bank: &Arc<Bank>,
+    ) {
+        match transaction_committed_status {
+            Some(transaction_committed_status) => Self::update_transaction_costs(
+                transaction_costs,
+                transaction_qos_results,
+                transaction_committed_status,
+                bank,
+            ),
+            None => {
+                Self::remove_transaction_costs(transaction_costs, transaction_qos_results, bank)
+            }
+        }
+    }
+
+    fn update_transaction_costs<'a>(
+        transaction_costs: impl Iterator<Item = &'a TransactionCost>,
+        transaction_qos_results: impl Iterator<Item = &'a transaction::Result<()>>,
+        transaction_committed_status: &Vec<bool>,
         bank: &Arc<Bank>,
     ) {
         let mut cost_tracker = bank.write_cost_tracker().unwrap();
         transaction_costs
             .zip(transaction_qos_results)
```

Comment:
let's create an iterator that filters `transaction_qos_results` by `transaction_committed_status` here, then pass it along with `transaction_costs` through to `remove_transaction_costs()` and avoid the replicode in the `and_then()`

Reply:
Can use some help on this one. How do I create an iterator of `transaction_qos_results` where its corresponding `transaction_committed_status` is false, that can still be zipped with `transaction_costs`? If I use `filter()` then the result iterator may not be the same length as `transaction_costs`; if use `map()`, do I have to clone a new set of `Result<>`s?

Reply:
something like this should do it
```rust
let uncommited_qos_results = transaction_qos_results
    .zip(transaction_committed_status)
    .filter_map(|(qos_result, committed)| (!*committed).then(|| qos_result)));
remove_transaction_costs(transactions_costs, uncommitted_qos_results, bank);
```

Reply:
With `filter_map`, wouldn't `uncommited_qos_results` have less items if some of `transaction_committed_status` is `true`, therefore it can't be `zip` with `transactions_costs`?